src/app/Library/CrudPanel/Traits/Read.php

```diff
@@ -203,31 +207,97 @@ public function getDefaultPageLength()
-     *                    or 2d array (first array: page length values, second array: page length labels)
-     *                    More at: https://datatables.net/reference/option/lengthMenu
+     * @param array|int $menu
+     *
+     * https://backpackforlaravel.com/docs/4.1/crud-cheat-sheet#page-length
      */
     public function setPageLengthMenu($menu)
     {
+        if (is_array($menu)) {
+            // start checking $menu integrity
+            if (count($menu) !== count($menu, COUNT_RECURSIVE)) {
+                // developer defined as setPageLengthMenu([[50, 100, 300]]) or setPageLengthMenu([[50, 100, 300],['f','h','t']])
+                // we will apply the same labels as the values to the menu if developer didn't
+                if (! in_array(0, $menu[0])) {
+                    if (! isset($menu[1]) || ! is_array($menu[1])) {
+                        $menu[1] = $menu[0];
+                    }
+                } else {
+                    abort(500, 'You should not use 0 as a key in paginator. If you are looking for "ALL" option, use -1 instead.');
+                }
+            } else {
+                // developer defined setPageLengthMenu([10 => 'f', 100 => 'h', 300 => 't']) OR setPageLengthMenu([50, 100, 300])
+                $menu = $this->buildPageLengthMenuFromArray($menu);
+            }
+        } else {
+            // developer added only a single value setPageLengthMenu(10)
+            if ($menu !== 0) {
+                $menu = [[$menu], [$menu]];
+            } else {
+                abort(500, 'You should not use 0 as a key in paginator. If you are looking for "ALL" option, use -1 instead.');
+            }
+        }
+
         $this->setOperationSetting('pageLengthMenu', $menu);
     }
 
+    /**
+     * Builds the menu from the given array. It works out with two different types of arrays:
+     *  [1, 2, 3] AND [1 => 'one', 2 => 'two', 3 => 'three'].
+     *
+     * @param array $menu
+     * @return array
+     */
+    private function buildPageLengthMenuFromArray($menu)
+    {
+        // check if the values of the array are strings, in case developer defined:
+        // setPageLengthMenu([0 => 'f', 100 => 'h', 300 => 't'])
+        if (count(array_filter(array_values($menu), 'is_string')) > 0) {
+            $values = array_keys($menu);
+            $labels = array_values($menu);
+
+            if (! in_array(0, $values)) {
+                return [$values, $labels];
+            }
+
+            abort(500, 'You should not use 0 as a key in paginator. If you are looking for "ALL" option, use -1 instead.');
+        } else {
+            // developer defined length as setPageLengthMenu([50, 100, 300])
+            // we will use the same values as labels
+            if (! in_array(0, $menu)) {
+                return [$menu, $menu];
+            }
+
+            abort(500, 'You should not use 0 as a key in paginator. If you are looking for "ALL" option, use -1 instead.');
```

Comment:
```suggestion
            if (in_array(0, $menu)) {
                abort(500, 'You should not use 0 as a key in paginator. If you are looking for "ALL" option, use -1 instead.');
            }

            return [$menu, $menu];
```

Reply:
I am not sure we should validate for 0 before the actual code. 

I think we should build it as `the most expected scenario` to be at first. And it's totally unexpected that someone will use 0. 

@tabacitu can you help us to decide here ?

Reply:
Gladly buuuut... I don't understand the issue sorry 😂 Both your original code and @promatik 's suggestion look fine to me - the way I understand it is we try to see if `0` is an array key - if so we throw a nasty error to let the dev know he shouldn't do that.

I'd prefer returning early like @promatik proposed.

Reply:
Oh yeah now I get it 😂 The question is whether to:
- (a) assume it doesn't have zero
- (b) assume it might have zero

I get what you're saying now @pxpm . There's small performance difference there, but for the sake of following the "return early" principle, we can use @promatik 's suggestion this time. It's easier to read the functions if you assume their structure is exceptions first, then at the very end what happens in the best case scenario (the end-game, if you will).

Reply:
I think in terms of performance I believe it is the same, because we are also validation it is **different** from zero, or zero is **not** in array.

In my perspective it is easier to read the code in this case when you do "nasty" code at the beginning;
_If there's a zero abort!!! Ok then, do everything that should be done ..._

And in the future, if we ever come to realize we should add zero support, or zero safe fallback, or even remove this abort for whatever reason, it's easy to just remove this lines, to commit and to review it also.

But again, this is an opinionated suggestion 😅